src/psyclone/domain/lfric/kernel_arg_order.py

```diff
@@ -0,0 +1,1063 @@
+            # Provide arguments common to LMA operators and fields on
+            # a space *unless* this is an inter-grid or CMA
+            # matrix-matrix kernel
+            if metadata.kernel_type not in [
+                    "cma-matrix-matrix", "inter-grid"]:
+                self.fs_common(function_space)
+
+            # Provide additional arguments if there is a
+            # field on this space
+            if [arg for arg in metadata.meta_args_get(
+                    [FieldArgMetadata, FieldVectorArgMetadata])
+                    if arg.function_space == function_space]:
+                self.fs_compulsory_field(function_space)
+
+            # Provide additional arguments if there is a
+            # intergrid field on this space
+            intergrid_field = [arg for arg in metadata.meta_args_get(
+                [InterGridArgMetadata, InterGridVectorArgMetadata])
+                               if arg.function_space == function_space]
+            if intergrid_field:
+                self.fs_intergrid(intergrid_field[0])
+
+            cma_ops = [arg for arg in metadata.meta_args_get(
+                ColumnwiseOperatorArgMetadata) if function_space in [
+                    arg.function_space_to, arg.function_space_from]]
+            if cma_ops:
+                if metadata.kernel_type == "cma-assembly":
+                    # CMA-assembly requires banded dofmaps
+                    self.banded_dofmap(function_space, cma_ops[0])
+                elif metadata.kernel_type == "cma-apply":
+                    # Applying a CMA operator requires indirection dofmaps
+                    self.indirection_dofmap(
+                        function_space, cma_ops[0])
+
+            # Provide any optional arguments. These arguments are
+            # associated with the keyword arguments (basis function
+            # and differential basis function) for a function space.
+            meta_funcs = metadata.meta_funcs \
+                if metadata.meta_funcs else []
+            if [func for func in meta_funcs if func.basis_function and
+                    func.function_space == function_space]:
+                self.basis(function_space)
+            if [func for func in meta_funcs if func.diff_basis_function
+                    and func.function_space == function_space]:
+                self.diff_basis(function_space)
+
+        # The boundary condition kernel (enforce_bc_kernel) is a
+        # special case.
+        if metadata.procedure_name and \
+                metadata.procedure_name.lower() == "enforce_bc_code":
+            self.field_bcs_kernel()
+
+        # The operator boundary condition kernel
+        # (enforce_operator_bc_kernel) is a special case.
+        if metadata.procedure_name and \
+                metadata.procedure_name.lower() == "enforce_operator_bc_code":
+            self.operator_bcs_kernel()
+
+        # Reference-element properties
+        if metadata.meta_ref_element:
```

Comment:
I could, but the assumption is that if a method is called then the associated arguments are required. I think it is simpler that way, rather than having to add checks in each, especially as there will be more than one subclass.